Wip

NikitaMatckevich opened the pull request:
Discussion of WIP  branch

ssizova committed:
refactoring

NikitaMatckevich committed:
dijkstra algorithm

Dijkstra algo implementation. Non-oriented graph is an upper-diagonal matrix of doubles.

ssizova commented on dijkstra.h:
```diff
@@ -0,0 +1,31 @@
+  vector<double> costs_;
+  id_t n_;
+public:
+  no_graph(vector<double> && costs, id_t n)
```
что тут означает двойной &&?

NikitaMatckevich replied:
Пока у нас нет никакого понятия как удобнее будет конструировать матрицу из геометрических объектов, я решил создать временный простой конструктор. Мне удобно внешне задать вектор длин, а потом мувнуть его внутрь матрицы.

NikitaMatckevich replied:
https://isocpp.org/blog/2012/10/classics-file-c-rvalue-references-explained
Вот топовая статья Томаса Беккера о rvalue ссылках, если интересно

ssizova commented on dijkstra.cpp:
```diff
@@ -0,0 +1,53 @@
+
+list<id_t> dijkstra(id_t s, id_t t, no_graph const& g) {
+  auto n = g.num_vertices();
+  vector<double> path_l(n, inf);
```
почему тут бесконечность?

ssizova replied:
Зачем мы сначала сами же создаем path_l, а потом shrink_to_fit? Я не очень это поняла :)

NikitaMatckevich replied:
Ну а как? Это вроде бы самый простой способ, не надо думать о топологии вершин, считать соседей и тп. За надежность этой бесконечности отвечает стандартная библиотека).
Если мы создаем вектор размера n, под него обычно выделяется больше памяти, чем его size. Потому что иначе любой pushback (даже один единственный) сразу бы заставил нас копировать весь вектор в другое место. shrink_to_fit осовобождает лишнюю память, потому что у нас ни одного пушбэка не будет.

ssizova replied:
С бесконечностью поняла.
Но Шринк, по-моему, имеет смысл использовать только после resize, потому что он может приводить к реалокации и лишнему выделению/освобождению памяти. Если ты не хочешь его динамически использовать, то можно просто вместо вектора взять обычный array. Ну или без шринка

NikitaMatckevich replied:
Можно и array. shrink_to_fit надо использовать всегда, когда ты не хочешь лишней capacity, а не только после resize.

ssizova committed:
corrected errors

ssizova committed:
Add basic classes

NikitaMatckevich committed:
Non-oriented graphs added

NikitaMatckevich committed:
Merge branch 'wip' of https://github.com/ssizova/Trajectory into wip

NikitaMatckevich committed:
Changed dijkstra file organisation

NikitaMatckevich committed:
Added function for reading config file

NikitaMatckevich committed:
Added new STL including header

NikitaMatckevich committed:
Errors fixed

ssizova committed:
small changes

ssizova committed:
added tests

ssizova committed:
Added tests. You can run them only if you have compiled gtest library on your laptop

ssizova committed:
changed itersection

anastasiyaDulepova committed:
Added obstacle

anastasiyaDulepova committed:
Added graph construction

anastasiyaDulepova committed:
Create config.txt

NikitaMatckevich committed:
changed tabs and const references

NikitaMatckevich committed:
fixed bug arc orientation

NikitaMatckevich committed:
moved file IO to std::filesystem

NikitaMatckevich committed:
filesystem crossplatform handlers

NikitaMatckevich committed:
changed tabs and const references

NikitaMatckevich committed:
fixed bugs graph construction

NikitaMatckevich committed:
simple script

NikitaMatckevich committed:
reading of map and SFINAE tricks

Reading of the map is now a part of config file.
User now can choose any IO files on his system to be used by the program.
This code uses std::filesystem library so should work fine on any OS.
Code is tolerant to the changes of containers used to store map elements. SFINAE complile-time tricks are implemented to ensure correct container behavior.

NikitaMatckevich committed:
work with graphs included in fucntion

Now the function that was initially used only to fill the graph matrice manipulates with the graphs independently, and calls dijkstra function inside. The resulting value is an STL container with pointers to map elements

NikitaMatckevich committed:
Noexcept added

Some little bugs fixed, added noexcept optimization

NikitaMatckevich committed:
Added final optimization

Word final enables some compilers to cut the road on virtual calls

NikitaMatckevich committed:
User interface/script added

One can use program by writing a script like that. Possible modification: switch in the main function on the desired graph type

NikitaMatckevich committed:
Json settings added

NikitaMatckevich committed:
Added my CMake

To build the program you need something like that. However I didn't want to delete Sofya's gtest environment

NikitaMatckevich committed:
Get rid of unused files

NikitaMatckevich committed:
Delete config.txt

Finally chosen another name for 1000 time

NikitaMatckevich committed:
IO files added

These files should be in your filesystem::path::current_path() folder if you don't want to provide optional arguments for your program. input.txt is to be written by user. outout.txt will be updated automatically, just put any txt file with such name it in your folder

NikitaMatckevich committed:
Delete STL.h as unused

NikitaMatckevich committed:
Remove unused files

NikitaMatckevich committed:
Less indirections, should now move little faster

Too many copies to deal with, but should work fine and still more efficient than pointer-based topologies

anastasiyaDulepova committed:
Before merge

anastasiyaDulepova committed:
Merge remote-tracking branch 'origin/wip' into wip

ssizova committed:
Add tests

anastasiyaDulepova committed:
Error with correct segment fixed

anastasiyaDulepova committed:
Merge branch 'wip' of https://github.com/ssizova/Trajectory into wip

anastasiyaDulepova committed:
delete out/build dir

anastasiyaDulepova committed:
Octave visualisation code
